src/mame/altos/altos586_hdc.cpp

```diff
@@ -0,0 +1,371 @@
+
+    [2] Preliminary 8600 User Manual
+        Section 7. System Specs.
+	4.5. Rigid Disk Controllers and Interface
+	Specification Revision 4.2, May 27, 1982, Page 52-62
+
+    [3] Preliminary 8600 User Manual
+        Section 7. System Specs.
+	A. Rigid Disk Controller Documentation
+	Specification Revision 4.2, May 27, 1982, Page 95-108
+
+    Note: The Altos 8600 manuals [2] and [3] describe a controller
+    wired differently (the 8089 seems to be in local mode, not remote),
+    the register addresses are different, but their function and command
+    set seems very different. I've written this before discovering those
+    manuals, so the command names are slightly different. My intent is to
+    eventually align that and review some of my assumptions too.
+
+    TODO: How is the command completion interrupt signalled?
+    I've only verified it to work with the diags floppy and that
+    one is not interrupt-driven. XENIX almost certainly is.
+
+***************************************************************************/
+
+#include "emu.h"
+#include "altos586_hdc.h"
+
+//#define VERBOSE 1
+#include "logmacro.h"
+
+DEFINE_DEVICE_TYPE(ALTOS586_HDC, altos586_hdc_device, "altos586_hdc", "Disk Controller board for Altos 586")
+
+altos586_hdc_device::altos586_hdc_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, ALTOS586_HDC, tag, owner, clock)
+	, m_bus(*this, finder_base::DUMMY_TAG, -1)
+	, m_iop(*this, "iop")
+	, m_hdd(*this, "hdd%u", 0)
+{
+}
+
+bool altos586_hdc_device::sector_exists(uint8_t index)
+{
+	if (!m_geom[m_drive]) {
+		logerror("drive not present: %d", m_drive);
+	} else if (m_head > m_geom[m_drive]->heads) {
+		logerror("head %d not present in drive that has %d heads", m_head, m_geom[m_drive]->heads);
+	} else if (index > m_geom[m_drive]->sectors) {
+		logerror("sector %d not present in drive that has %d sectors per track", index, m_geom[m_drive]->sectors);
+	} else {
+		LOG("drive %d sector CHS=%d/%d/%d found\n", m_drive, m_cyl[m_drive], m_head, index);
+		return true;
+	}
+
+	m_status |= 0x40; // Record not found.
+	return false;
+}
+
+uint32_t altos586_hdc_device::sector_lba(uint8_t const index)
+{
+	return (m_cyl[m_drive] * m_geom[m_drive]->heads + m_head) * m_geom[m_drive]->sectors + index;
+}
+
+void altos586_hdc_device::sector_read(uint8_t index)
+{
+	if (!sector_exists(index))
+		return;
+	m_hdd[m_drive]->read(sector_lba(index), &m_sector[5]);
+	m_iop->drq1_w(ASSERT_LINE);
+}
+
+void altos586_hdc_device::sector_write(uint8_t index)
+{
+	if (!sector_exists(index))
+		return;
+	m_hdd[m_drive]->write(sector_lba(index), &m_sector[5]);
+}
+
+uint16_t altos586_hdc_device::mem_r(offs_t offset, uint16_t mem_mask)
+{
+	return m_bus->read_word_unaligned(offset << 1, mem_mask);
+}
+
+void altos586_hdc_device::mem_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	m_bus->write_word_unaligned(offset << 1, data, mem_mask);
+}
```

Comment:
Can you just use `read_word` and `write_word` here?  You’ll only get aligned memory accesses through the address space read/write handlers, so it shouldn’t be necessary to use the ones that have extra code for unaligned accesses.